execution-engine/src/wasi/wasmtime.rs

```diff
@@ -0,0 +1,720 @@
+            Ok(o) => o,
+        }
+    };
+}
+
+/// Impl the MemoryHandler for Caller.
+/// This allows passing the Caller to WASIWrapper on any VFS call.
+impl<'a> MemoryHandler for Caller<'a> {
+    fn write_buffer(&mut self, address: u32, buffer: &[u8]) -> ErrNo {
+        let memory = match self
+            .get_export(WASIWrapper::LINEAR_MEMORY_NAME)
+            .and_then(|export| export.into_memory())
+        {
+            Some(s) => s,
+            None => return ErrNo::NoMem,
+        };
+        let address = address as usize;
+        unsafe {
+            std::slice::from_raw_parts_mut(memory.data_ptr().add(address), buffer.len())
+                .copy_from_slice(buffer)
+        };
+        ErrNo::Success
+    }
+
+    fn read_buffer(&self, address: u32, length: u32) -> Result<Vec<u8>, ErrNo> {
+        let length = length as usize;
+        let memory = match self
+            .get_export(WASIWrapper::LINEAR_MEMORY_NAME)
+            .and_then(|export| export.into_memory())
+        {
+            Some(s) => s,
+            None => return Err(ErrNo::NoMem),
+        };
+        let mut bytes: Vec<u8> = vec![0; length];
+        unsafe {
+            bytes.copy_from_slice(std::slice::from_raw_parts(
+                memory.data_ptr().add(address as usize),
+                length,
+            ))
+        };
+        Ok(bytes)
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Checking function well-formedness.
+////////////////////////////////////////////////////////////////////////////////
+
+/// Checks whether `main` was declared with `argc` and `argv` or without in the
+/// WASM program.
+fn check_main(tau: &ExternType) -> EntrySignature {
+    match tau {
+        ExternType::Func(tau) => {
+            let params = tau.params();
+
+            if params == &[ValType::I32, ValType::I32] {
+                EntrySignature::ArgvAndArgc
+            } else if params == &[] {
+                EntrySignature::NoParameters
+            } else {
+                EntrySignature::NoEntryFound
+            }
+        }
+        _otherwise => EntrySignature::NoEntryFound,
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// The Wasmtime host provisioning state.
+////////////////////////////////////////////////////////////////////////////////
+/// The facade of WASMTIME host provisioning state.
+pub struct WasmtimeRuntimeState {}
+
+////////////////////////////////////////////////////////////////////////////////
+// Operations on the WasmtimeRuntimeState.
+////////////////////////////////////////////////////////////////////////////////
+
+impl WasmtimeRuntimeState {
+    /// Creates a new initial `HostProvisioningState`.
+    pub fn new(filesystem: Arc<Mutex<FileSystem>>, program_name: &str) -> Self {
+        // Load the VFS ref to the global environment. This is required by Wasmtime.
+        *VFS_INSTANCE.lock().unwrap() =
```

Comment:
Should this not be written in terms of the locking macro?  If not, can we replace the `unwrap` with `expect`?

Reply:
I add `?`.